### DISCUSSION TITLE
cytoscape.js based widget not rendering

### DISCUSSION THREAD

**Discussion by anmolbhatia05** [manzt/anywidget#627] [JavaScript] [answered]
Hello all, 

I am trying to create a widget using anywidget and [cytoscape.js](https://js.cytoscape.org/) in a vs code jupter notebook env. My aim is to create a graph visualiser and editor for my thesis.

The following code wont render anything when the cell is executed. 

```py
import anywidget
import traitlets


class Widget(anywidget.AnyWidget):
    _esm = """
    import cytoscape from "https://cdn.jsdelivr.net/npm/cytoscape@3.30.1/+esm";

    function render({ model, el }) {
      let cy = cytoscape({
        elements: [ // list of graph elements to start with
          { data: { id: "a" } }, // node a
          { data: { id: "b" } }, // node b
          { data: { id: "ab", source: "a", target: "b" } } // edge ab
        ],
        style: [ // the stylesheet for the graph
          {
            selector: "node",
            style: {
              "background-color": "#666",
              "label": "data(id)"
            }
          },
          {
            selector: "edge",
            style: {
              "width": 3,
              "line-color": "#ccc",
              "target-arrow-color": "#ccc",
              "target-arrow-shape": "triangle",
              "curve-style": "bezier"
            }
          }
        ],
        layout: {
          name: "grid",
          rows: 1
        }
      });
      cy.mount(el);
      console.log(el);
      console.log(cy);
    }

    export default { render };
    """
    _css="""
    """

w = Widget()
w
```

This is the output of the cell 
<img width="830" alt="Screenshot 2024-07-24 at 17 10 38" src="https://github.com/user-attachments/assets/494ebcec-398b-47a2-bf09-2492df958847">

The dev tools output also seems fine - 
<img width="537" alt="Screenshot 2024-07-24 at 17 18 03" src="https://github.com/user-attachments/assets/51ed281b-c810-4de1-93f6-0e5284e073d1">
 
 Any advice/directions/critique is welcome!

**Comment by manzt** [comment, marked as answer]
Thanks for sharing your project. It sounds exciting!

Your issue seems related to integrating cytoscape.js rather than anywidget itself. I'd like to avoid adding noise to the anywidget issues with challenges that aren't related to anywidget functionality (hence I've transferred this to a Discussion). For help with this integration in the future, I'd recommend posting in GitHub Discussions or our [Discord](https://discord.gg/W5h4vPMbDQ).

That said, a quick glimpse at [the documentation](https://js.cytoscape.org/#getting-started) has a very distinct disclaimer:

> Note that Cytoscape.js uses the dimensions of your HTML DOM element container for layouts and rendering at initialisation. Thus, it is very important to place your CSS stylesheets in the <head> before any Cytoscape.js-related code. Otherwise, dimensions may be sporadically reported incorrectly, resulting in undesired behaviour.

```py
import anywidget
import traitlets


class Widget(anywidget.AnyWidget):
    _esm = """
    import cytoscape from "https://cdn.jsdelivr.net/npm/cytoscape@3.30.1/+esm";

    function render({ model, el }) {
        let container = document.createElement("div");
        container.style.height = "300px";
        el.appendChild(container)
        
        let cy = cytoscape({
          container: container,
          elements: [ // list of graph elements to start with
            {  data: { id: 'a' } },
            { data: { id: 'b' } },
            { data: { id: 'ab', source: 'a', target: 'b' } }
          ],
          style: [ // the stylesheet for the graph
            {
              selector: 'node',
              style: {
                'background-color': '#666',
                'label': 'data(id)'
              }
            },
            {
              selector: 'edge',
              style: {
                'width': 3,
                'line-color': '#ccc',
                'target-arrow-color': '#ccc',
                'target-arrow-shape': 'triangle',
                'curve-style': 'bezier'
              }
            }
          ],
          layout: { name: 'grid', rows: 1 }
        });
    }

    export default { render };
    """

w = Widget()
w
```

The issue with your example is that your container (`el`) doesn't have a height.

**Comment by ianhi** [comment]
@anmolbhatia05 you may find it helpful to make use of the code in the ipycytoscape repository: https://github.com/cytoscape/ipycytoscape?tab=readme-ov-file#ipycytoscape

or if you are interested take up maintership of that package including but not limited to transitioning it to use anywidget
